### PR TITLE
Adds batch set/delete operations to Properties and a static Create function for Environment

### PR DESCRIPTION
**Pull request by a-type**
The implementation may not be 100% intuitive, nor 100% elegant, which is unfortunate. Here are some design decisions which need an opinion:

* `setAll` / `deleteAll` return a batch `restore` function; i.e. calling the `restore` function will undo all modifications from `setAll` or `deleteAll`. Batch `restore`s are also pushed on the internal `change` stack.

* `Environment.create` does a batch delete of keys in `process.env` before continuing, which feels a little odd to me, but then again, I don't see another way to clear the environment.

Another note on the existing codebase: should the `"restoring multiple single-value changes"` test in `Properties_spec.js` (line `308` in my branch) include `foo` in the initial Properties constructor hash to really test deletion restoration?

I took the liberty of adding some test coverage for cases that interested me and didn't seem covered: calling an individually created `restore` function from a set/delete operation and then proceeding to call the full `restore` on the `Properties` instance.

### CODE REVIEW

**Comment by jagoda** [comment]
> calling the restore function will undo all modifications from setAll or deleteAll. Batch restores are also pushed on the internal change stack.

This sounds reasonable to me.

> Environment.create does a batch delete of keys in process.env before continuing, which feels a little odd to me, but then again, I don't see another way to clear the environment.

I will comment on this inline when I get to it (hard to follow from the description).

> Another note on the existing codebase: should the "restoring multiple single-value changes" test in Properties_spec.js (line 308 in my branch) include foo in the initial Properties constructor hash to really test deletion restoration?

Will come back to this one.

> I took the liberty of adding some test coverage for cases that interested me and didn't seem covered

Sounds like a good idea to me!

**Comment by jagoda** [comment]
> Another note on the existing codebase: should the "restoring multiple single-value changes" test in Properties_spec.js (line 308 in my branch) include foo in the initial Properties constructor hash to really test deletion restoration?

The intent of that case is to make sure that the environment is restored to it's original state when doing multiple different single operations (doesn't necessarily matter what they are). Delete is an interesting edge case because `process.env.FOO = undefined` results in `process.env.FOO === "undefined"`. I believe the specific mechanics of the delete should be covered in the other test cases. Am I missing something?

**Comment by jagoda** [comment]
Looks like there is a global leak failure on Node 0.11. We should probably address this in order to merge the PR.

**Comment by a-type** [comment]
> Am I missing something?

I don't think so. I just had some trouble understanding the idea behind the test.

Also, I don't know what's up with the globals. I don't recognize any of them. The same globals fail the test on `master` as well. I'm assuming it's a change to a dependency or Node 0.11.

**Comment by jagoda** [comment]
My bet is on Node 0.11. I would recommend checking the issue tracker for [lab](http://github.com/hapijs/lab).

**Comment by a-type** [comment]
Looks like it was fixed in Lab 5.0.0, but that version also removed the assertion features from Lab and put them in Code. I tried using `Code.expect` but it's missing some features, strangely.

**Comment by a-type** [comment]
After a deeper look at their docs, looks like they're switching convention from `to.have.property("a", 0)` to `to.include({ a : 0 })`.

Do you want to move to Code or use a different library?

**Comment by jagoda** [comment]
In general I have moved away from Lab because it has lots of little oddities and they weren't very willing to work with me on some of them. In this case though, I would say just go with whatever is quickest/easiest (migration to another library could be more work than it's worth right now).

**Comment by a-type** [comment]
I threw together a branch with Lab 5.x.x and Code. It wasn't too hard with such a small codebase. I'll put a PR for that up.

**Comment by jagoda** [comment]
Could we get a rebase of this on `master` so that the diff doesn't include the changes from #3?

**Comment by a-type** [comment]
Done.

**Comment by jagoda** [comment]
Cool, I think this is good. Will publish a new version momentarily.

**Comment by jagoda** [comment]
v0.3.0
